### PR TITLE
feat: refactor valid placement method and place method

### PR DESCRIPTION
**Pull request by Gabe-Torres**
I added the overlap method as a helper method for the valid placement method. I refactored the place method to store the coordinates of a ship placed into an "existing coordinates" array that is set as empty in the initialize. The data grabbed in the place method is used in the overlap method. I also added some refactors to the valid placement method on lines 29 and 31. Setting the consecutive cood and diagonal coord to == false and == true, which gives it something to compare.

### CODE REVIEW

File: lib/board.rb [hunk 1]
```diff
@@ -50,8 +57,17 @@ def diagonal_coordinates?(coordinates)
...
+    @existing_coordinates += coordinates
     coordinates.each do |coordinate|
       @cells[coordinate].place_ship(ship)
     end
+      @existing_coordinates
+  end
+  
+
+  def overlap?(coordinates)
+    coordinates.any? do |coord|
```
Comment (Gabe-Torres) [inline lib/board.rb:69]: .any? is able to iterate over two arrays and look for overlap, then return true.

File: lib/board.rb [hunk 2]
```diff
@@ -21,20 +22,26 @@ def valid_coordinate?(coordinate)
...
       false
-    else
+    else 
       true
     end
   end
 
   def consecutive_coordinates?(coordinates)
-    letters = coordinates.map { |coordinate| coordinate[0].ord }
-    numbers = coordinates.map { |coordinate| coordinate[1..-1].to_i }
```
Comment (Gabe-Torres) [inline lib/board.rb:37]: I changed this for readability for myself while going over our code, and forgot to change it back

File: lib/board.rb [hunk 3]
```diff
@@ -50,8 +57,17 @@ def diagonal_coordinates?(coordinates)
   end
 
   def place(ship, coordinates)
+    @existing_coordinates += coordinates
```
Comment (Gabe-Torres) [inline lib/board.rb:60]: decided to update the place method to use within overlap. We should make some tests to make sure everything runs smoothly but it didn't affect any other test. += combined the coordinate array with the existing coordinates array.

File: lib/board.rb [hunk 4]
```diff
@@ -21,20 +22,26 @@ def valid_coordinate?(coordinate)
...
   def valid_placement?(ship, coordinates)
-    if coordinates.length != ship.length
+    if overlap?(coordinates) == true
+      false
+    elsif coordinates.length != ship.length
       false
     elsif consecutive_coordinates?(coordinates) == false
       false
-    elsif diagonal_coordinates?(coordinates)
+    elsif diagonal_coordinates?(coordinates) == true
```
Comment (Gabe-Torres) [inline lib/board.rb:31]: Needed something to compare the method to or it would always return false regardless of the outcome.

File: lib/board.rb [hunk 5]
```diff
@@ -21,20 +22,26 @@ def valid_coordinate?(coordinate)
   end
 
   def valid_placement?(ship, coordinates)
-    if coordinates.length != ship.length
+    if overlap?(coordinates) == true
```
Comment (Gabe-Torres) [inline lib/board.rb:25]: put the overlap method at the beginning. I think it could work anywhere within the method

File: lib/board.rb [hunk 6]
```diff
@@ -1,15 +1,16 @@
 class Board
-  attr_reader :cells
-
+  attr_reader :cells,
+              :existing_coordinates
   def initialize
     @cells = make_cells
+    @existing_coordinates = []
```
Comment (Gabe-Torres) [inline lib/board.rb:6]: added the array to store coordinates within for valid placement method

File: lib/board.rb (hunk 1 above)
Comment (DChavez18) [reply lib/board.rb:69]: So thankful to find that .any? method

File: lib/board.rb (hunk 3 above)
Comment (DChavez18) [reply lib/board.rb:60]: Looks great! I like the update on the place method!

File: lib/board.rb (hunk 4 above)
Comment (DChavez18) [reply lib/board.rb:31]: Nice! I know we had it working but it did feel incomplete!

File: lib/board.rb (hunk 5 above)
Comment (DChavez18) [reply lib/board.rb:25]: Yeah, placement shouldn't be an issue with methods in classes, because they are only read when called upon.

**Review by DChavez18 — APPROVED** [review]

**Pull request merged**